Railsアプリを追加

a5ob7r opened the pull request:
# 概要

APIサーバのためのRailsアプリを追加する。

およそ以下のような作業を行なった。

- APIモードでの新規Railsアプリの追加
- Dcoker化
- Rubocopの追加
- RSpecの追加
- CI用のジョブの追加

a5ob7r committed:
Initialize the Rails app

Initialize it using a following command line.

```sh
rails new . --skip-bundle --api --minimal --database=postgresql
```

a5ob7r committed:
Add rubocop

a5ob7r committed:
Add RSpec

a5ob7r committed:
Fix database configs

a5ob7r committed:
Add database schemas

a5ob7r committed:
Add how to setup to README

a5ob7r committed:
Remove unnecessary directories

a5ob7r committed:
Add how to run test to README

a5ob7r committed:
Add CI jobs

a5ob7r committed:
Add a platform to supports

a5ob7r committed:
Fix a code in accordance with Rubocop

a5ob7r committed:
Fix the hostname of the database on CI

a5ob7r committed:
Add a config to check the DB health

a5ob7r committed:
Add some environment variables to configure the DB

a5ob7r committed:
Add a missing RSpec runner spec

a5ob7r committed:
Bump puma to 6.3.0

tk0miya commented on .env.sample:
雑談: うちの会社だとデータベースには MySQL を使うことが多いですね。Rails から見たときは意識することは少ないので、今の時点ではあまり気にしなくてもよさそうではあるけど、情報共有です。

※ 特化した機能を使う場合、チューニングする場合などは RDBMS エンジンがなにかを気にすることになります。

tk0miya commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,66 @@
+    services:
+      db:
+        image: postgres:15.3-bullseye
+        env:
+          POSTGRES_PASSWORD: password
+          POSTGRES_USER: root
```
質問: 単に興味なんですが、ここってもう一度宣言必要なんでしたっけ? ひとつ上の階層で宣言しているので、ここでの定義が必要なのかちょっと気になりました。

a5ob7r replied:
それを期待して、最初はこの記述を書いていなかったんですけど、どうやら `jobs.<job_id>.env` (ひとつ上の階層の宣言) の値はここには引き継がれないようでした。このコンテナでは `POSTGRES_PASSWORD` の宣言が必須なのですが、それが宣言されていないといった旨のエラーが発生しました。
https://github.com/sls-training/2023-API-training-server/actions/runs/5139914386/jobs/9250857448

ホストマシン上での設定とコンテナ上の設定ということでコンテキストは異なるので、この仕様は納得ではあるけど、階層的にはしっくりこない気もする。

a5ob7r replied:
そういえば、最上位のフィールドにも `env` があるけど、こっちに設定したら引き継がれたりするのかな。

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#env

a5ob7r replied:
あとで試してみよう。

tk0miya replied:
あら、そうなんですね。実験済みなら問題ないです。resolve にします。

a5ob7r replied:
> そういえば、最上位のフィールドにも `env` があるけど、こっちに設定したら引き継がれたりするのかな。
> 
> https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#env

#18 で試してみましたが、こちらもできないようでした。

a5ob7r committed:
Remove comment-outed gems

tk0miya commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,66 @@
+
+    services:
+      db:
+        image: postgres:15.3-bullseye
```
ここってバージョン固定必要ですか?

* バージョンを固定する
    * ソフトウェアが変化しないので、システムが壊れない (安定する)
    * 不具合やセキュリティホールが見つかっても解消されない
    * 古いバージョンはサポートを受けられない可能性がある
    * 不具合の存在の継続的な調査や定期的なアップデートを行う必要がある
* バージョンを固定しない
    * ソフトウェアの更新によってシステムが壊れることがある
    * 不具合やセキュリティホールはいずれ修正される (期待できる)

バージョンを固定しないことによるデメリットもあるので、固定してはいけないとまでは言えないのですが、固定した場合はバージョンアップを計画しておく必要があります。

ここに関して言うと、

* データベースの機能は比較的安定しているので、バージョンを固定する意味はあまりないのでは
* 固定するとしても 15.3 ではなく 15 でよいのでは
* bullseye と OS まで指定する必要はあるのか

あたりが気になりました。


ちなみに DB だけでなく、ライブラリやフレームワーク (Gemfile や package.json で管理しているもの) のバージョンアップについても考える必要がありますね。
※ ライブラリ類は .lock ファイルもコミットしてバージョン固定をするのが一般的なので、定期的なアップデートを検討する必要がありますね。

a5ob7r replied:
バージョンを固定しないことで、自動的にいい感じにバージョンアップしてくれればいいのですが、一般にそうではないと思いますので、それなら明示的に指定しておき、定期的なバージョンアップを考えたほうがいいのでは、と考えています。幸い、ライブラリであれば `dependabot` のような自動化の仕組みがありますし、他のものに似たようなものがあるのではと思っています。

まあ、今回はトレーニングなので、バージョンを固定しないことによるリスクも無視していいとは思いますが。

tk0miya replied:
ツールや設定ファイルによって自動的にバージョンアップされるもの、されないものがありますが、ここについては指定次第で自動的にバージョンアップされますよね。
`postgres:latest` とか `postgres:15` と指定すると、15.3 → 15.4 → 15.5 ... と実行の際に条件にあう最新版を使うようになりますよね (latest の場合は `16.x` も使ってくれるはず)。

※ だから、バージョン固定を解除しようよ、とまで突っ込みたいわけではないですが…

KT-tim replied:
安全にいくなら固定するのがいいと思いますが、固定するなら定期的にdbのバージョンを変える必要があるか調査するメンテナンスタスクが必要になるかもしれません

tk0miya commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,66 @@
+      - name: Run Rubocop
+        run: bundle exec rubocop
+
+  actionlint:
```
actionlint しらなかった。へー
https://rhysd.hatenablog.com/entry/2021/07/11/214313

tk0miya commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+    postgresql-client \
+    imagemagick \
+    git \
+    ; \
```
この最後の行、ちょっと違和感があるんだけど、こういうもんです?

a5ob7r replied:
`;` のことですかね。

tk0miya replied:
はい、そうです。

a5ob7r replied:
発想としては配列やハッシュの Trailing Comma に近いですね。こうしておくと、インストールしたいパッケージを追加するときに `;` の位置を調整せずに済む、という考えです。

```Dockerfile
RUN set -eux; \
  apt-get install -y --no-install-recommends \
    git \
    ; \
```

のとき、

```Dockerfile
RUN set -eux; \
  apt-get install -y --no-install-recommends \
    git \
    tig \
    ; \
```

とできるけど、

```Dockerfile
RUN set -eux; \
  apt-get install -y --no-install-recommends \
    git; \
```

のときは、

```Dockerfile
RUN set -eux; \
  apt-get install -y --no-install-recommends \
    git \
    tig; \
```

となってしまう、といったところです。詰まるところ、宗教の話ですね。

tk0miya replied:
一般的に(?) Dockerfile でのコマンドの区切りには `&&` を使います。
https://docs.docker.jp/engine/articles/dockerfile_best-practice.html#run
`;` を使うと前のコマンドが失敗してもそのまま次に進んでしまうため、 `&&` を使っているという解釈です。

そして `&&` を使う場合は、末尾につけるということはできないはずなので、こういう書き方はしないですね。

※ `set` コマンドは失敗しないので今回はツッコミは入れなかったのですが、別のコマンドが `;` でぶら下がっている場合はツッコミいれてました


ちなみに、docs.docker.jp では行頭に `&&` をつけるスタイルで書かれていますね。
僕もこのスタイルで書くことが多いです。

```
RUN mkdir -p /usr/src/things \
    && curl -SL http://example.com/big.tar.xz \
    | tar -xJC /usr/src/things \
    && make -C /usr/src/things all
```

a5ob7r replied:
僕も以前はそのスタイルを利用していたのですが、今はコマンドを実行する前に、 `set -e` を設定することによる早期終了を利用しています。

まあ、どっちでもいいとは思いますが、強いていうなら、後者の方が各コマンドのインデントを揃えやすいので気に入っています。

確認してみたら、RubyのオフィシャルのDockerfileはこのスタイルで書かれていますね。
https://github.com/docker-library/ruby/blob/4041b21f3e11111846e6b6043da2da92e1da7019/3.2/bullseye/Dockerfile

ついでにHaskellでも。
https://github.com/haskell/docker-haskell/blob/6a4bd2d6ac51a08af489476145b1f83c8d20e575/9.4/buster/Dockerfile

tk0miya replied:
ふむ、そういうスタイルとして捉えるのが良いかー。見慣れないのでもやっとしただけなので、狙いが分かって理解しました (納得はしてないw)。
ということで、このままで行きましょう。

tk0miya commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+    build-essential \
+    libpq-dev \
+    postgresql-client \
+    imagemagick \
```
よく分かってないけど、imagemagick 使うんです?

a5ob7r replied:
あー、これはRailsチュートリアルで利用していたものをそのままにしてしまっていますね。削除しておきます。

tk0miya commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+ruby-3.2.2
```
これってこの書き方であってましたっけ? 手元だとバージョン名だけ書いてあるものが多いのだけど、最近書き方変わったのかな。

a5ob7r replied:
`rails new` による自動生成の結果ですね。僕も違和感はありました。でも実装名があったほうが明示的でいいのかな。省略されるとruby(MRI)扱いなんですかね。

tk0miya replied:
Rails が生成したやつならきっとあってるんだろうなあ

僕の知る限りでは

* このファイルは rbenv という Ruby 切り替えツールが参照するファイル
* rbenv は MRI のバージョンを数字のみで表現している (他の実装は prefix がつく)

というものなので、ruby- という prefix が付いてるのは初めて見ました。

まあ、動かしてみて問題があったら考えればいいと思います。
今回は docker compose 経由で (ruby イメージ経由で) 動かすので、rbenv は関わらず、問題も起きない気がしますし。

a5ob7r replied:
今回はrbenvは利用する想定ではないですし、このファイル自体はアプリケーションの実行に不要なので、削除しようかと思っていたのですが、 `ruby/setup-ruby` がこのファイルを読んで自動で用意するRubyのバージョンを決定してくれるようなので残してあります。

https://github.com/ruby/setup-ruby#supported-version-syntax

tk0miya replied:
slack に書いたけど、以下のテンプレートから生成されているみたいですね。
https://github.com/rails/rails/blob/d026af43c35c25814b4d16511f1b8fc6d70a4569/railties/lib/rails/generators/rails/app/templates/ruby-version.tt

おそらく rbenv が有効でない環境で生成するとこれが出てくるんだと思います。

`ruby/setup-ruby` が使っているのは知らなかったのだけど、これで問題なさそうなので、このスレッドは resolve にします。

tk0miya commented on Gemfile:
```diff
@@ -0,0 +1,61 @@
+  gem 'rubocop', '~> 1.51', require: false
+  gem 'rubocop-performance', '~> 1.18', require: false
+  gem 'rubocop-rails', '~> 2.19.1', require: false
+  gem 'rubocop-rspec', '~> 2.22', require: false
```
つい最近、 rubocop-rspec から factory_bot 関連の cop が rubocop-factory_bot という別パッケージに切り出されました。
https://github.com/rubocop/rubocop-rspec/blob/master/CHANGELOG.md#2220-2023-05-06

ですから、rubocop-factory_bot も参照するようにしてください。

tk0miya commented on compose.yaml:
雑談: compose.yaml を使うのが推奨になっていたのね… しらなかった。

tk0miya commented on compose.yaml:
```diff
@@ -0,0 +1,37 @@
+x-app-template: &app-template
+  image: rails_on_docker_2023-api-training-server:0.0.0
```
感想: バージョンって更新予定あります? 付ける意味なさそうだなあと思いました

a5ob7r replied:
過去に、バージョン指定をつけておかないと、 `x-hoge` の記法 でイメージを共通化させた際に、ビルド時に同じイメージを同時にビルドしてしまうということがありました。それだと、せっかくのイメージの利点がなくなってしまうので、わざとしてあります。

tk0miya replied:
へー、知らなかった

tk0miya commented on compose.yaml:
```diff
@@ -0,0 +1,37 @@
+      POSTGRES_USER:
+    volumes:
+      - "db-data:${PGDATA:-/var/lib/postgresql/data}"
+    expose:
+      - "5432"
```
postgresql イメージはデフォルトで 5432 番ポートを expose しているので、わざわざ書かなくて大丈夫ですよ
https://github.com/docker-library/postgres/blob/a23c0e97980edae5be2cd4eb68ff1f0762d031cd/15/bullseye/Dockerfile#L218

a5ob7r replied:
本当ですね。Railsチュートリアルの時はこれがないと接続できなかったので書いていました。勘違いだったのかな。削除しておきます。

tk0miya commented on config/database.yml:
```diff
@@ -0,0 +1,94 @@
+  #
+  # GitHub Actions上でのCI実行時には、環境変数 CI=true が常に設定されている。
+  # https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
+  host: <%= ENV['CI'] == 'true' ? 'localhost' : 'db' %>
```
この手のものをハードコーディングすると、環境が増えたときに厄介なので、環境変数で差し込めるようにするのがよいですね。
```suggestion
  host: <%= ENV['POSTGRES_HOST'] %>
```

環境変数を省略した場合にデフォルト値を与えたい、って場合はこう書くかな。
```suggestion
  host: <%= ENV.fetch('POSTGRES_HOST', 'localhost') %>
```

a5ob7r replied:
確かに単純に環境変数経由で値を設定するようにしたほうがシンプルでいいですね。修正します。

tk0miya replied:
実際のプロジェクトでは、いろんな環境を用意することになるので、外から指定できるようになっていると便利です

* ローカルの開発環境
* 本番環境
* その他の環境 (ステージング環境、QA環境などなど)

tk0miya commented on .gitignore:
```diff
@@ -0,0 +1,17 @@
+# Ignore bundler config.
+/.bundle
+
+# Ignore all logfiles and tempfiles.
+/log/*
+/tmp/*
```
雑談: 本番用の Docker イメージを作る場合、log や tmp ディレクトリがないと Rails サーバが起動できずにクラッシュするのでご注意を。Dockerfile で定義するか、もともと書いてあったように .keep を残しておくかの対策が必要です。

a5ob7r replied:
あー、 Dockerイメージを作ることは考えていませんでした。今は、 `bin/setup` 内で実行される `bin/rails log:clear tmp:clear` のタスクで作成されるので、わざわざリポジトリに含めなくてもいいだろうという判断でした。

~Dockerイメージだけにイメージしていませんでしたってか、ガハハ。~

tk0miya reviewed (commented):
作業ごとにコミットが分かれているのはいいですね。
Rails が生成した部分と、自分で書き換えた部分の違いがわかるのでレビューしやすくて助かりました。

ひとつお願いがあるとすると `rails new` したときにどういうオプションを付けたのかを残しておくと最高です。全部眺めたわけではないけど、今回ってなにかオプション指定してません?


あと、もうひとつアドバイスすると、まったく関係ないコミットの場合、過去の PR にコミットをぶら下げないほうがマージが早くなる可能性があります。
たとえば、今回の場合は openapi.yml の変更とは一切関係ないので、 `rails new` しました、という PR を投げると追い越してマージが可能です。
ブランチ管理がちょっとややこしくなるものの、マージはそれなりに時間がかかるものなので、追い越してマージしやすくするという戦略は自分の助けになるはず。

※ そういう面では API 定義についても、分割すると楽になるものはいくつかありそうですね。

a5ob7r commented:
> ひとつお願いがあるとすると rails new したときにどういうオプションを付けたのかを残しておくと最高です。全部眺めたわけではないけど、今回ってなにかオプション指定してません?

今回は以下のようなオプションを利用しました。

```sh
rails new . --skip-bundle --api --minimal --database=postgresql
```

このPRはまだWIPという建前なので、リベースしてRailsを作成したコミットのコミットメッセージにこのオプションを含めておきます。

a5ob7r commented:
> あと、もうひとつアドバイスすると、まったく関係ないコミットの場合、過去の PR にコミットをぶら下げないほうがマージが早くなる可能性があります。
> たとえば、今回の場合は openapi.yml の変更とは一切関係ないので、 rails new しました、という PR を投げると追い越してマージが可能です。
> ブランチ管理がちょっとややこしくなるものの、マージはそれなりに時間がかかるものなので、追い越してマージしやすくするという戦略は自分の助けになるはず。
>
> ※ そういう面では API 定義についても、分割すると楽になるものはいくつかありそうですね。

これまでリベース前提で考えていたので、できるだけリニアになるように作業しようとしていたところはありますね。そこへのこだわりは、少なくとも私用以外では捨てようと思っているので、これから意識していきたいです。

a5ob7r committed:
Add rubocop-factory_bot explicitly to Gemfile

a5ob7r committed:
Remove an unnecessary port expose

It is exposed by default.

a5ob7r committed:
Use ENV and fetch() to set the database hostname

This's very simple and no need to write service specific configs.

a5ob7r committed:
Remove an unnecessary package from Dockerfile

a5ob7r commented:
リベースして、親ブランチをこの時点での main である 4b301cb0d9edb89d7abdfb57a891ca5f94c56405 に変更しました。

tk0miya commented on compose.yaml:
```diff
@@ -0,0 +1,36 @@
+    ports:
+      - "${APP_PORT:-3000}:3000"
+    environment:
+      POSTGRES_HOST:
```
ここは env で指定するのではなく、compose.yml で固定しちゃっていいんじゃないかなあ。
```suggestion
      POSTGRES_HOST: db
```

a5ob7r replied:
まあ、コンテナを利用する際にホスト名を変更したいというシチュエーションはまずないから、ハードコードしてしまっていいのかな。

a5ob7r replied:
でも、いざ変更したいとなったときに、 `compose.yaml` に値を指定してあると、上書きするためには、優先度的にコマンドライン上(環境変数 or 引数)で直接指定するしかないみたいです。それをやるくらいなら、他の環境変数と同様に `.env` 経由で指定できるようにしておいたほうがラクなのでは、と思います。

https://docs.docker.com/compose/environment-variables/envvars-precedence/

tk0miya replied:
docker compose を使っていて (db コンテナを起動するようにしておいて)、他の値に変更したいというシチュエーションがあまり思いつかないですね。
柔軟性があるように見えて、その実特に柔軟ではないというのはあまり意味がないのでは。
設定項目が増えることによって余計なトラブルを引き起こすだけでは?

プロジェクトの全員が Linux/UNIX であるとかミドルウェアに強いというわけではなく、プログラムは書けるが環境については得意ではないというメンバーは存在します。
(スキルは人それぞれ偏っているので、Rails は得意ではないが〜〜が得意というのは普通にありえる)

そうしたときに、使わない設定項目を増やすことはマイナスの価値を生むというのも知っておくと良さそうです。
具体的なメリットのあるシチュエーションがあるのであれば、追加するのには反対ではないのですが、いまのところ思いつかないので、うーん 🤔 という印象です。

※ まあ、-1 するほどではないかな、というところなので、これがいいという意見であれば引き下がるていどのものなので、このコメントで最後にします。

a5ob7r replied:
なるほど、これもある種の認知負荷の問題なんですかね。確かによく変更する設定とほとんど変更しない設定が並んでいると、人によっては混乱の元になりそうですね。

というか、今回のDBのホスト名の場合は、環境変数だけ変更しても、結局 `compose.yaml` に記載されているDBコンテナのラベル名も変更しないとアクセスできないから、なおさら意味がなかったですね。

というわけで修正します。

tk0miya replied:
はい、そのとおりですね

* db コンテナの名前を変えるのであれば、compose.yml を書き換えるはず
    * このときに .env を書き換えるという話?
    * ファイルが分かれている価値はありそう?
* db コンテナ以外のデーターベースに接続したい
    * ほんとにそういうことやる?

ということで、柔軟性が上がっているとはいえないという認識でした。

ですから「ほとんど変更しない設定」ではなく「まったく変更しない設定」という理解でした。

tk0miya replied:
ここまで書いて気づいたけど、データベースの ID, パスワードも変えたいというニーズはなさそうな気がしました。
セキュアにする必要がある場所ではないですし、compose.yml に固定値を書いてもよさそう

※ MySQL では `MYSQL_ALLOW_EMPTY_PASSWORD: 'yes'` という設定をして、root ユーザのパスワード空にするアプローチをよくやりますね

a5ob7r replied:
こうしておくと、 `.env` を書き換えるだけで複数のコンテナの環境変数設定を同時に変更できるので、環境構築時のように試行錯誤しているときは便利なんですよね。でも、それが完了した後だと、確かに変更したいという状況はなかなかないのかな。

なくてもいいならないほうがいいの精神でこちらも修正しておきます。

tk0miya reviewed (approved):
ひとつコメント追加したけど、おおよそ問題なさそうなのでフライング +1 です。

a5ob7r committed:
Add the default value of POSTGRES_HOST

This value isn't probably changed by users.

a5ob7r committed:
Add default values for POSTGRES_USER and POSTGRES_PASSWORD

Nobody probably needs to change these values after building development
environment.

a5ob7r committed:
Remove .env

We no longer use it, especially for docker compose.
